### PR TITLE
Define and implement transaction

### LINKED ISSUE #58: implement things like `Do.afterCommit` in gryphon

### PR DESCRIPTION
**Pull request by bellatoris**
* isolationLevel 변경가능한 transaction manager 구현

fix #58

### CODE REVIEW

**Comment by bellatoris** [comment]
Current dependencies on/for this PR:
* main
  * **PR #55** <a href="https://app.graphite.dev/github/pr/Stair-Crusher-Club/scc-server/55" target="_blank">Graphite</a> 
    * **PR #56** <a href="https://app.graphite.dev/github/pr/Stair-Crusher-Club/scc-server/56" target="_blank">Graphite</a>  👈
      * **PR #57** <a href="https://app.graphite.dev/github/pr/Stair-Crusher-Club/scc-server/57" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Stair-Crusher-Club/scc-server/56?utm_source=stack-comment).

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/Transaction.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package club.staircrusher.infra.persistence.sqldelight
+
+import app.cash.sqldelight.TransactionWithReturn
+import club.staircrusher.stdlib.persistence.Transaction
+
+class Transaction<T>(
```
Comment (Zeniuus) [inline subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/Transaction.kt:6]:
```suggestion
class SqlDelightTransaction<T>(
```
그냥 Transaction하면 넘 헷갈릴 듯 ㅜ

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package club.staircrusher.infra.persistence.sqldelight
+
+import app.cash.sqldelight.Query
+import app.cash.sqldelight.driver.jdbc.JdbcDriver
+import java.sql.Connection
+import javax.sql.DataSource
+
+class JdbcDriver(
```
Comment (Zeniuus) [inline subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt:8]: 얘도 SqlDelightJdbcDriver
코드 읽는데 이게 sqldelight 내부의 JdbcDriver인지 우리가 정의한 건지 넘 헷갈리네유

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/DB.kt [hunk 3]
```diff
@@ -32,4 +36,38 @@ class DB(dataSource: DataSource) {
...
+
+    override fun <T> doInTransaction(block: Transaction<T>.() -> T): T {
+        val driver = this.driver as JdbcDriver
+        check(driver.isolationLevel == null) {
+            """
+            Since SCC does not allow nested transaction, isolationLevel saved in
+            thread local must be null.
+            """.trimIndent()
+        }
+        return transactionWithResult(noEnclosing = true) {
```
Comment (Zeniuus) [inline subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/DB.kt:48]: 이중 트랜잭션 무조건 방지? 나쁘지 않아

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/DB.kt [hunk 4]
```diff
@@ -32,4 +36,38 @@ class DB(dataSource: DataSource) {
...
+            Transaction(this).block()
+        }
+    }
```
Comment (Zeniuus) [inline subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/DB.kt:51]: 직접 구현 대신 fun <T> doInTransaction(isolationLevel: TransactionIsolationLevel, block: Transaction<T>.() -> T)를 호출하면 어때요?

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    private val dataSource: DataSource,
+): JdbcDriver() {
+    private val _isolationLevel = ThreadLocal<Int>()
+    var isolationLevel: Int?
```
Comment (Zeniuus) [inline subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt:12]:
```suggestion
    var oldIsolationLevel: Int?
```
어때요?

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    override fun addListener(listener: Query.Listener, queryKeys: Array<String>) {
+        // No-op. JDBC Driver is not set up for observing queries by default.
+    }
+
+    override fun removeListener(listener: Query.Listener, queryKeys: Array<String>) {
+        // No-op. JDBC Driver is not set up for observing queries by default.
+    }
+
+    override fun notifyListeners(queryKeys: Array<String>) {
+        // No-op. JDBC Driver is not set up for observing queries by default.
```
Comment (Zeniuus) [inline subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt:65]: 구현 안 하는 메소드들은 super 거 불러도 되지 않을까용

**Review by Zeniuus — APPROVED** [review]

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/Transaction.kt (hunk 1 above)
Comment (bellatoris) [reply subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/Transaction.kt:6]: 근데 얘는 구현체니까 거의 볼일이 없지 않아요?

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/DB.kt (hunk 4 above)
Comment (bellatoris) [reply subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/DB.kt:51]: 그러면 무조건 connection에 isloationLevel을 설정하는게 날아가서 성능이 안좋아질 것 같아요

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt (hunk 5 above)
Comment (bellatoris) [reply subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt:12]: 이 값이 꼭 old 값만을 들고 있는 것은 아니긴 합니다.

File: subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt (hunk 6 above)
Comment (bellatoris) [reply subprojects/persistence_model/src/main/kotlin/club/staircrusher/infra/persistence/sqldelight/JdbcDriver.kt:65]: 이건 abstract method를 구현한거라서 super가 없어용

**Comment by bellatoris** [comment]
**Graphite Merge Job**

Current status: ⏳ Queued to merge

This pull request is currently queued to be merged as part of a stack.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Stair-Crusher-Club/scc-server/56?utm_source=stack-comment).


Job Reference: Pv7IckxFsqqlUNIROeMA

**Comment by bellatoris** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Stair-Crusher-Club/scc-server/56?utm_source=stack-comment).


Job Reference: GPrmpXx1RQXYQgyKqvrw

**Pull request merged**